File: 프로그래머스/2/147354. 테이블 해시 함수/테이블 해시 함수.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <cmath>

using namespace std;

int cmp_idx = -1;

bool cmp(const vector<int>& a, const vector<int>& b) {
    if (a[cmp_idx] != b[cmp_idx])
        return a[cmp_idx] < b[cmp_idx];
    else
        return a[0] > b[0];
}

int get_val(const vector<int>& d, int i) {
    int val = 0;
    for (const auto& n : d)
        val += (n % i);
    
    return val;
}

string change_binary(int n) {
    string result;
    
    while (n > 0) {
        result.insert(result.begin(), '0' + (n % 2));
        n /= 2;
    }
    
    return result.empty() ? "0" : result;
}

string xor_op(const string& a, const string& b) {
    string long_s = (a.size() > b.size()) ? a : b;
    string short_s = (a.size() <= b.size()) ? a : b;
    
    // Padding short_s with leading zeros
    short_s.insert(short_s.begin(), long_s.size() - short_s.size(), '0');
    
    for (size_t i = 0; i < short_s.size(); ++i) {
        long_s[i] = (long_s[i] != short_s[i]) ? '1' : '0';
    }
    
    return long_s;
}

int solution(vector<vector<int>> data, int col, int row_begin, int row_end) {
    int answer = 0;
    
    cmp_idx = col - 1;
    sort(data.begin(), data.end(), cmp);
    
    // 3번 과정
    vector<int> result;
    for (int i = row_begin - 1; i <= row_end - 1; ++i)
        result.push_back(get_val(data[i], i + 1));
    
    // XOR
    if (result.size() == 1)
        return result[0];
    
    string prev = change_binary(result[0]);
    
    for (size_t i = 1; i < result.size(); ++i) {
        string r = change_binary(result[i]);
        prev = xor_op(prev, r);
    }
    
    // 계산
    for (size_t i = 0; i < prev.size(); ++i)
        if (prev[i] == '1')
            answer += (1 << (prev.size() - 1 - i));
    
    return answer;
}

File: 프로그래머스/2/147354. 테이블 해시 함수/테이블 해시 함수.java
import java.util.*;
import java.util.stream.*;

class Solution {
    public int solution(int[][] data, int col, int row_begin, int row_end) {
        // 데이터 정렬
        Arrays.sort(data, (a, b) -> {
            if (a[col - 1] == b[col - 1]) return b[0] - a[0];
            return a[col - 1] - b[col - 1];
        });

        // 각 행에 대한 계산
        List<Integer> list = IntStream.range(1, data.length + 1)
            .map(idx -> Arrays.stream(data[idx - 1])
                .reduce(0, (a, b) -> a + (b % idx)))
            .boxed()
            .collect(Collectors.toList());

        // 주어진 범위의 행들에 대해 XOR 연산 수행
        int result = list.get(row_begin - 1);
        for (int i = row_begin; i < row_end; i++) {
            result ^= list.get(i);
        }

        return result;
    }
}